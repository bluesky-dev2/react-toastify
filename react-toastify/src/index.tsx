export { useToastContainer } from './hooks';
export { cssTransition } from './utils';
export { ToastContainer, Bounce, Flip, Slide, Zoom } from './components';
export { toast } from './core';
export * from "./types"