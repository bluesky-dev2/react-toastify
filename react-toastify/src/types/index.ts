import * as React from 'react';

export type ToastPosition =
  | 'top-right'
  | 'top-center'
  | 'top-left'
  | 'bottom-right'
  | 'bottom-center'
  | 'bottom-left';

export type TypeOptions = 'info' | 'success' | 'warning' | 'error' | 'default';
export type ToastContent = React.ReactNode | (() => React.ReactNode );
export interface Toast {
  content: ToastContent;
  options: InternalToastOptions;
}

type Id = number | string
export type ToastId = Id;
export type ContainerId = Id;

interface CommonOptions {
  /**
   * Pause the timer when the mouse hover the toast.
   * `Default: true`
   */
  pauseOnHover?: boolean;

  /**
   * Pause the toast when the window loose focus.
   * `Default: true`
   */
  pauseOnFocusLoss?: boolean;

  /**
   * Remove the toast when clicked.
   * `Default: true`
   */
  closeOnClick?: boolean;

  /**
   * Set the delay in ms to close the toast automatically.
   * Use `false` to prevent the toast from closing.
   * `Default: 5000`
   */
  autoClose?: number | false;

  /**
   * Set the default position to use.
   * `One of: 'top-right', 'top-center', 'top-left', 'bottom-right', 'bottom-center', 'bottom-left'`
   * `Default: 'top-right'`
   */
  position?: ToastPosition;

  /**
   * Pass a custom close button.
   * To remove the close button pass `false`
   */
  closeButton?: React.ReactNode | boolean;

  /**
   * An optional css class to set for the progress bar.
   */
  progressClassName?: string | object | null;

  /**
   * An optional style to set for the progress bar.
   */
  progressStyle?: object;

  /**
   * An optional css class to set.
   */
  className?: string | object | null;

  /**
   * An optional css class to set for the toast content.
   */
  bodyClassName?: string | object | null;

  /**
   * Hide or show the progress bar.
   * `Default: false`
   */
  hideProgressBar?: boolean;

  /**
   * Pass a custom transition built with react-transition-group.
   */
  transition?: React.ComponentType;

  /**
   * Allow toast to be draggable
   * `Default: true`
   */
  draggable?: boolean;

  /**
   * The percentage of the toast's width it takes for a drag to dismiss a toast
   * `Default: 80`
   */
  draggablePercent?: number;

  /**
   * Define the ARIA role for the toast
   * `Default: alert`
   *  https://www.w3.org/WAI/PF/aria/roles
   */
  role?: string;

  /**
   * Set id to handle multiple container
   */
  containerId?: ContainerId;

  /**
   * Fired when clicking inside toaster
   */
  onClick?: Function;

  /**
   * Support right to left display.
   * `Default: false`
   */
  rtl?: boolean;
}

export interface ToastOptions extends CommonOptions {
  /**
   * Called when toast is mounted.
   */
  onOpen?: () => void;

  /**
   * Called when toast is unmounted.
   */
  onClose?: () => void;

  /**
   * Set the toast type.
   * `One of: 'info', 'success', 'warning', 'error', 'default'`
   */
  type?: TypeOptions;

  /**
   * Set a custom `toastId`
   */
  toastId?: ToastId;

  /**
   * Used during update
   */
  updateId?: ToastId;

  /**
   * Set the percentage for the controlled progress bar. `Value must be between 0 and 1.`
   */
  progress?: number | string;

  /**
   * Add a delay in ms before the toast appear.
   */
  delay?: number;
}

export interface InternalToastOptions extends ToastOptions {
  staleId?: ToastId;
  toastId: ToastId;
  key: Id;
  closeToast: () => void;
  position: ToastPosition;
}

export interface UpdateOptions extends ToastOptions {
  /**
   * Used to update a toast.
   * Pass any valid ReactNode(string, number, component)
   */
  render?: ToastContent;
}

export interface ToastContainerProps extends CommonOptions {
  /**
   * Whether or not to display the newest toast on top.
   * `Default: false`
   */
  newestOnTop?: boolean;

  /**
   * An optional inline style to apply.
   */
  style?: object;

  /**
   * An optional css class for the toast.
   */
  toastClassName?: string | object | null;

  /**
   * Show the toast only if it includes containerId and it's the same as containerId
   * `Default: false`
   */
  enableMultiContainer?: boolean;
}

// interface Toast {
//   /**
//    * Shorthand to display toast of type 'success'.
//    */
//   success(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'info'.
//    */
//   info(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'warning'.
//    */
//   warn(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'error'.
//    */
//   error(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Check if a toast is active by passing the `toastId`.
//    * Each time you display a toast you receive a `toastId`.
//    */
//   isActive(toastId: ToastId): boolean;

//   /**
//    * Remove a toast. If no `toastId` is used, all the active toast
//    * will be removed.
//    */
//   dismiss(toastId?: ToastId): void;

//   /**
//    * Update an existing toast. By default, we keep the initial content and options of the toast.
//    */
//   update(toastId: ToastId, options?: UpdateOptions): void;

//   /**
//    * Listen for change when a toast is added or removed. The number of toast displayed is passed as paran to the callback
//    */
//   onChange(
//     callback: (count?: number, containerId?: string | number) => void
//   ): void;

//   /**
//    * Set a controlled progress bar value to 100% then close the toast
//    */
//   done(toastId: ToastId): void;

//   /**
//    * Let you define `ToastContainer` props when lazy mounted.
//    * When called enable lazy mounted container
//    */
//   configure(config?: ToastContainerProps): void;

//   /**
//    * Display a toast without a specific type.
//    */
//   (content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Helper to set notification type
//    */
//   TYPE: Type;

//   /**
//    * Helper to set position
//    */
//   POSITION: Position;
// }

// /**
//  * Helper to set notification type
//  */

// /**
//  * Helper to set position
//  */

// /**
//  * Helper to build custom entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */
