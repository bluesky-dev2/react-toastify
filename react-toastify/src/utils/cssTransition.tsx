import React from 'react';
import { Transition } from 'react-transition-group';
import { TransitionProps } from 'react-transition-group/Transition';

import { NOOP } from './constant';

export interface CSSTransitionProps {
  /**
   * Css class to apply when toast enter
   */
  enter: string;

  /**
   * Css class to apply when toast leave
   */
  exit: string;

  /**
   * Define the duration of the transition in ms
   * You can also pass an array `[enterDuration, exitDuration]`
   * `Default: 750`
   */
  duration?: number | [number, number];

  /**
   * Append current toast position to the classname.
   * For instance `myclass--top-center`...
   * `Default: false`
   */
  appendPosition?: boolean;
}

interface AnimationProps extends Partial<TransitionProps> {
  position?: Position;
  preventExitTransition?: boolean;
  
}

export function cssTransition({
  enter,
  exit,
  duration = 750,
  appendPosition = false
}: CSSTransitionProps) {
  const Animation: React.FC<AnimationProps> = ({
    children,
    position,
    preventExitTransition,
    ...props
  }) => {
    const enterClassName = appendPosition ? `${enter}--${position}` : enter;
    const exitClassName = appendPosition ? `${exit}--${position}` : exit;
    let enterDuration: number, exitDuration: number;

    if (Array.isArray(duration) && duration.length === 2) {
      [enterDuration, exitDuration] = duration;
    } else {
      enterDuration = exitDuration = duration as number;
    }

    const onEnter = (node: HTMLElement) => {
      node.classList.add(enterClassName);
      node.style.animationFillMode = 'forwards';
      node.style.animationDuration = `${enterDuration * 0.001}s`;
    };
    const onEntered = (node: HTMLElement) => {
      node.classList.remove(enterClassName);
      node.style.cssText = '';
    };
    const onExit = (node: HTMLElement) => {
      node.classList.add(exitClassName);
      node.style.animationFillMode = 'forwards';
      node.style.animationDuration = `${exitDuration * 0.001}s`;
    };

    return (
      <Transition
        {...props}
        timeout={
          preventExitTransition
            ? 0
            : {
                enter: enterDuration,
                exit: exitDuration
              }
        }
        onEnter={onEnter}
        onEntered={onEntered}
        onExit={preventExitTransition ? NOOP : onExit}
      >
        {children}
      </Transition>
    );
  };
  return Animation;
}
