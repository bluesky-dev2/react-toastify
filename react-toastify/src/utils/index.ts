export * from './propValidator';
export * from './eventManager';
export * from './constant';
export * from './cssTransition';
