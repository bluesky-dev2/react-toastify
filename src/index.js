import ToastContainer from './ToastContainer';
import toastify from './toastify';

export { ToastContainer, toastify as toast };
