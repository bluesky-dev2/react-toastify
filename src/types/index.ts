import * as React from 'react';

export type ToastPosition =
  | 'top-right'
  | 'top-center'
  | 'top-left'
  | 'bottom-right'
  | 'bottom-center'
  | 'bottom-left';

export type TypeOptions = 'info' | 'success' | 'warning' | 'error' | 'default';
export type ToastContent = React.ReactNode | ((props?: any) => React.ReactNode);
export interface Toast {
  content: ToastContent;
  options: WithInjectedOptions;
}

type Id = number | string;
type ClassName = string | null;
export type ToastId = Id;
export type ContainerId = Id;

interface CommonOptions {
  /**
   * Pause the timer when the mouse hover the toast.
   * `Default: true`
   */
  pauseOnHover?: boolean;

  /**
   * Pause the toast when the window loose focus.
   * `Default: true`
   */
  pauseOnFocusLoss?: boolean;

  /**
   * Remove the toast when clicked.
   * `Default: true`
   */
  closeOnClick?: boolean;

  /**
   * Set the delay in ms to close the toast automatically.
   * Use `false` to prevent the toast from closing.
   * `Default: 5000`
   */
  autoClose?: number | false;

  /**
   * Set the default position to use.
   * `One of: 'top-right', 'top-center', 'top-left', 'bottom-right', 'bottom-center', 'bottom-left'`
   * `Default: 'top-right'`
   */
  position?: ToastPosition;

  /**
   * Pass a custom close button.
   * To remove the close button pass `false`
   */
  closeButton?:
    | React.ReactElement
    | ((props: any) => React.ReactElement)
    | boolean;

  /**
   * An optional css class to set for the progress bar.
   */
  progressClassName?: ClassName;

  /**
   * An optional style to set for the progress bar.
   */
  progressStyle?: React.CSSProperties;

  /**
   * An optional css class to set.
   */
  className?: ClassName;

  /**
   * An optional css class to set for the toast content.
   */
  bodyClassName?: ClassName;

  /**
   * An optional inline style to apply for the toast content.
   */
  bodyStyle?: React.CSSProperties;

  /**
   * Hide or show the progress bar.
   * `Default: false`
   */
  hideProgressBar?: boolean;

  /**
   * Pass a custom transition built with react-transition-group.
   */
  transition?: React.ReactNode;

  /**
   * Allow toast to be draggable
   * `Default: true`
   */
  draggable?: boolean;

  /**
   * The percentage of the toast's width it takes for a drag to dismiss a toast
   * `Default: 80`
   */
  draggablePercent?: number;

  /**
   * Define the ARIA role for the toast
   * `Default: alert`
   *  https://www.w3.org/WAI/PF/aria/roles
   */
  role?: string;

  /**
   * Set id to handle multiple container
   */
  containerId?: ContainerId;

  /**
   * Fired when clicking inside toaster
   */
  onClick?: (event: React.MouseEvent) => void;

  /**
   * Support right to left display.
   * `Default: false`
   */
  rtl?: boolean;
}

export interface ToastOptions extends CommonOptions {
  /**
   * Called when toast is mounted.
   */
  onOpen?: <T = {}>(props: T) => void;

  /**
   * Called when toast is unmounted.
   */
  onClose?: <T = {}>(props: T) => void;

  /**
   * An optional inline style to apply.
   */
  style?: React.CSSProperties;

  /**
   * Set the toast type.
   * `One of: 'info', 'success', 'warning', 'error', 'default'`
   */
  type?: TypeOptions;

  /**
   * Set a custom `toastId`
   */
  toastId?: ToastId;

  /**
   * Used during update
   */
  updateId?: ToastId;

  /**
   * Set the percentage for the controlled progress bar. `Value must be between 0 and 1.`
   */
  progress?: number | string;

  /**
   * Add a delay in ms before the toast appear.
   */
  delay?: number;
}

export interface WithInjectedOptions extends ToastOptions {
  staleId?: ToastId;
  toastId: ToastId;
  key: Id;
  closeToast: () => void;
  position: ToastPosition;
  children?: ToastContent;
  draggablePercent: number;
  in?: boolean;
  progressClassName?: ClassName;
  className?: ClassName;
  bodyClassName?: ClassName;
  onExited?: () => void;
  // transition: Transition;
}

export interface UpdateOptions extends ToastOptions {
  /**
   * Used to update a toast.
   * Pass any valid ReactNode(string, number, component)
   */
  render?: ToastContent;
}

export interface ToastContainerProps extends CommonOptions {
  /**
   * Whether or not to display the newest toast on top.
   * `Default: false`
   */
  newestOnTop?: boolean;

  /**
   * An optional inline style to apply.
   */
  style?: React.CSSProperties;

  /**
   * An optional inline style to apply for the toast.
   */
  toastStyle?: React.CSSProperties;

  /**
   * An optional css class for the toast.
   */
  toastClassName?: ClassName;

  /**
   * Show the toast only if it includes containerId and it's the same as containerId
   * `Default: false`
   */
  enableMultiContainer?: boolean;

  /**
   * Limit the number of toast displayed at the same time
   */
  limit?: number;
}
