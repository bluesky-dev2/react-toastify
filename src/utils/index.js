export * from './constant';
export * from './propValidator';
export * from './eventManager';
export * from './cssTransition';
