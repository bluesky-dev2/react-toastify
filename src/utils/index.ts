export * from './propValidator';
export * from './constant';
export * from './cssTransition';
