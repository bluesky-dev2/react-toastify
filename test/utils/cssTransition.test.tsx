import * as React from 'react';
import { render } from '@testing-library/react';

import { cssTransition, POSITION } from '../../src/utils';
import { ToastTransitionProps } from '../../src/types';

const transitionProps: ToastTransitionProps = {
  done: jest.fn(),
  isIn: true,
  preventExitTransition: false,
  position: POSITION.TOP_RIGHT,
  nodeRef: {
    current: document.createElement('div')
  }
};

// TODO: write better tests ðŸ˜œ
describe('cssTransition helper', () => {
  it('Should return a valid react component', () => {
    const Transition = cssTransition({
      enter: 'foo',
      exit: 'bar'
    });

    const Child = () => <div>Plop</div>;
    const { container, getByText } = render(
      <Transition {...transitionProps}>
        <Child />
      </Transition>
    );

    expect(getByText('Plop').textContent).toContain('Plop');
    expect(container).toMatchSnapshot();
  });

  it('Should be possible to prevent exit transition', () => {
    const Transition = cssTransition({
      enter: 'foo',
      exit: 'bar'
    });

    const Child = () => <div>Plop</div>;
    const { container, getByText } = render(
      <Transition {...transitionProps} preventExitTransition>
        <Child />
      </Transition>
    );
    expect(getByText('Plop').textContent).toContain('Plop');
    expect(container).toMatchSnapshot();
  });

  it('Should be possible to disable collapse animation', () => {
    const Transition = cssTransition({
      enter: 'foo',
      exit: 'bar',
      collapse: false
    });

    const Child = () => <div>Plop</div>;
    const { container, getByText } = render(
      <Transition {...transitionProps}>
        <Child />
      </Transition>
    );
    expect(getByText('Plop').textContent).toContain('Plop');
    expect(container).toMatchSnapshot();
  });

  it('Should be possible to change  collapse duration', () => {
    const Transition = cssTransition({
      enter: 'foo',
      exit: 'bar',
      collapseDuration: 200
    });

    const Child = () => <div>Plop</div>;
    const { container, getByText } = render(
      <Transition {...transitionProps}>
        <Child />
      </Transition>
    );
    expect(getByText('Plop').textContent).toContain('Plop');
    expect(container).toMatchSnapshot();
  });
});
