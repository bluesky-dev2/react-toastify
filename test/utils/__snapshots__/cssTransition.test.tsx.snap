// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cssTransition helper Should be possible to change  collapse duration 1`] = `
<div>
  <div>
    Plop
  </div>
</div>
`;

exports[`cssTransition helper Should be possible to disable collapse animation 1`] = `
<div>
  <div>
    Plop
  </div>
</div>
`;

exports[`cssTransition helper Should be possible to prevent exit transition 1`] = `
<div>
  <div>
    Plop
  </div>
</div>
`;

exports[`cssTransition helper Should return a valid react component 1`] = `
<div>
  <div>
    Plop
  </div>
</div>
`;
